[UX] WIP: Improve gamepage

Kajot-dev opened the pull request:
This PR aims to improve UI/UX of the redesigned gamepage view

## Main fixes:
- Moved "report issues" link a bit so both panels will always be the same height
- Moved `TabPanel` into separate component as it was duplicated and gamepage used the same logic for the 3rd time
- Instead of buttons, used tabs with icons
  - Two new translations will be needed for tab headers
- Dots menu and game settings are on the right side back again
- Fixed `HowLongToBeat` not displaying properly
  - It will scale better with translations which have longer text (no more ellipses, but still not perfect)
  - It now uses MUI's Accordion
- Added initial styles for uninstall button (logic not implemented yet)
- Centered play button (see rationale)

## A little rationale for centered button
- Button was not resizable which caused visual glitches on some translations
- However I didn't want to go back to the old design where the button was stretched

## Some screenshots with comparison:

### Main info tab

**Before:**
![Screenshot_20240303_233535](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/17925111/9cc92c66-cbc4-43bc-8983-1e10a939cfd4)
**After:**
![Screenshot_20240303_233339](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/17925111/636ed4ff-8617-4fe4-b685-eb83a6ce6818)

### Extra info tab

**Before:**
![Screenshot_20240303_233459](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/17925111/a5c4b306-1394-4f2f-ae95-98cfd33b1ff4)
**After:**
![Screenshot_20240303_233401](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/17925111/cd004ba3-45ad-4aa5-8237-dd4088de5367)

### Sys requirements tab

**Before:**
![Screenshot_20240303_233523](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/17925111/ba0b2238-99c2-421b-a669-511af90fb5cb)
**After:**
![Screenshot_20240303_233435](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/17925111/21969f11-6705-423e-97ad-de67e095fc0d)

## TODO:

- [ ] Use more `--space-*` CSS variables for more consistency
- [ ] Implement uninstall button
- [ ] Make header (icons, back button) and footer (report issues) constant height so scrolling can be done by each panel independently

---

Use the following Checklist if you have changed something on the Backend or Frontend:

- [x] Tested the feature and it's working on a current and clean install.
- [x] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

flavioislima commented:
**CLA Assistant Lite bot:** <br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/main/doc/cla.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
   - - -
   I have read the CLA Document and I hereby sign the CLA
   - - -
   **Kajot-dev** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You can retrigger this bot by commenting **recheck** in this Pull Request</sub>

Kajot-dev commented:
I have read the CLA Document and I hereby sign the CLA

Kajot-dev commented:
- Top buttons have gained a shadow and are now sticky
- Anticheat warning does not have unnecessary margin and is scaled correctly now
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/17925111/68f49896-6d6c-4b3f-b4be-4d44432e6ddb)
- Removed min-height and increased padding to compensate in installinfo/extrainfo rows. Since height is now bounded by the padding, opening HLTB does not have this "jumping" effect
- Implemented Uninstall button

Redesigned gamepage should now be feature complete!

arielj commented on src/frontend/screens/Game/GamePage/components/HLTBInline.tsx:
```diff
@@ -0,0 +1,48 @@
+      <Accordion expanded={isExpanded} onChange={handleChange}>
+        <AccordionSummary
+          expandIcon={<ExpandMore />}
+          aria-control="hltb-content"
```
I get an invalid prop warning in the console with this

![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/188464/dea5b9bf-ba24-4831-95de-89a9e8ab5f09)

arielj commented on src/frontend/screens/Game/GamePage/components/HLTBInline.tsx:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  return (
+    <div className="hltbWrapper">
```
I'm not sure which component is problematic, but when I move using a controller (or keyboard), the item for the HLTB info doesn't have the outline (the background is a biiiit darker)

https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/188464/4c207812-e47e-450a-a23b-3b5edbd0db5e

Kajot-dev replied:
Good catch, it's probably MUI stuff :/

Kajot-dev replied:
Fixed, it was kinda tricky, because of MUI

arielj commented on src/frontend/components/UI/WikiGameInfo/components/HowLongToBeat/index.scss:
```diff
@@ -1,5 +1,7 @@
 .howLongToBeat {
```
some change here broke the style in the old design
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/188464/0e9185d2-8c5d-4b74-b8ac-e33c961b8a34)

Kajot-dev replied:
Fixed

arielj commented on src/frontend/screens/Game/GamePage/index.scss:
```diff
@@ -461,16 +480,42 @@
+    display: flex;
+    flex-direction: row;
+    gap: min(var(--space-lg), 50px);
+    position: sticky;
```
I think we should aim to not have to scroll this screen at all (so this being sticky shouldn't be needed). Scrolling is kinda annoying on the Steam Deck and when we have scroll in the whole screen it means the `play` button can start outside of the viewport when it makes sense for this button to always be visible since it's kinda the main action we want.

I think at some point I had that implemented (the other components would shrink if needed) but I can't see it working that way in the main branch either so I don't know what happened there.

If something like system requirements or game description are long we can have scroll on those elements that are not that critical for example.

Kajot-dev replied:
Main info element was *partially* implemented like this, so description would become scrollable, but `max-height` was too high and whole screen scrolled anyway. I will try to fix this.

Kajot-dev replied:
Should be fixed now. Now gamepage should almost always adapt to full screen and not be scrollable. One side note that under 600px gamepage *will* become scrollable since scrollable game description is not sufficient (for preventing overflow) anymore.

arielj commented on src/frontend/screens/Game/GamePage/index.scss:
```diff
@@ -461,16 +480,42 @@
+      }
+    }
+
+    .opts-left {
+      justify-content: flex-start;
+    }
+
+    .opts-right {
+      justify-content: flex-end;
+    }
```
these utility classes doesn't seem to be needed, the parent has `justify-content: space-between;` anyway

Kajot-dev replied:
Fixed

arielj commented on src/frontend/screens/Game/GamePage/index.scss:
```diff
@@ -461,16 +480,42 @@
+  .topRowWrapper {
+    display: flex;
+    flex-direction: row;
+    gap: min(var(--space-lg), 50px);
```
this gap is not really needed, we have only 2 children and they are separated on the 2 sides of the element

Kajot-dev replied:
Fixed

arielj commented on src/frontend/screens/Game/GamePage/index.scss:
```diff
@@ -461,16 +480,42 @@
 
+  .topRowWrapper {
+    display: flex;
+    flex-direction: row;
```
this is also not needed, row is the default anyway

Kajot-dev replied:
Fixed

arielj commented on src/frontend/screens/Game/GamePage/index.scss:
```diff
@@ -461,16 +480,42 @@
+    flex-direction: row;
+    gap: min(var(--space-lg), 50px);
+    position: sticky;
+    top: var(--space-lg);
```
this is not doing anything, the element is `position: static` so top has no effect

Kajot-dev replied:
It has the position `sticky`, not `static`, but since we don't want the whole screen to scroll (like you mentioned earlier), I will remove `sticky` and `top` entirely

arielj replied:
yeah, sorry, I was playing around and I remove the sticky incorrectly so it was just static for me when I checked this :see_no_evil:

arielj commented on src/frontend/screens/Game/GamePage/index.tsx:
```diff
@@ -416,17 +434,22 @@ export default React.memo(function GamePage(): JSX.Element | null {
+                    {!isBrowserGame && <SettingsButton gameInfo={gameInfo} />}
+                    <DotsMenu gameInfo={gameInfo} handleUpdate={handleUpdate} />
+                  </div>
+                </div>
```
a bit nit picky, but we can avoid using the `opts` wrappers if we have the 3 elements inside `topRowWrapper` and then you set `.topRowWrapper .backButton { margin-inline-end: auto }`

(you'd need to move the gap from `.opts` to `.topRowWrapper`)

Kajot-dev replied:
Fixed

arielj commented on src/frontend/screens/Game/GamePage/components/HLTBInline.tsx:
```diff
@@ -0,0 +1,48 @@
+import { Accordion, AccordionSummary, AccordionDetails } from '@mui/material'
+import HowLongToBeat from 'frontend/components/UI/WikiGameInfo/components/HowLongToBeat'
+
+const HLTBInline = () => {
```
maybe we can include the different logic from this inside the current `HLTB` component? and depending on the `experimentalFeatures.enableNewDesign` value? I did that already in the Scores component to make it inline

The reason is that at some point we are going to remove the old design and we'll have to remember to remove one of these HLTB components. If we just add some conditional code in the current HLTB component it's going to be easier to clean in the future.

Kajot-dev replied:
Good point! Only downside is that `isExpanded` will be present regardless of the design since react hook must be at component top level.

arielj replied:
yeah, not a big deal though, it's just a state that won't be used in the old design, but I wouldn't worry about that

Kajot-dev replied:
Fixed

arielj commented on src/frontend/screens/Game/GamePage/components/MainButton.tsx:
```diff
@@ -128,6 +142,13 @@ const MainButton = ({ gameInfo, handlePlay, handleInstall }: Props) => {
+  if (extraClassName) {
+    // do not override internal classNames
+    delete other.className
+  }
```
I understand this `className` prop is only used to add the `mainBtn` class, right? I'd rather just hardcode the `mainBtn` class inside this component since we are not going to use that for anything else

it will be added in the old design too, but it will do nothing there if there's no CSS anyway

Kajot-dev replied:
Fixed

Kajot-dev commented:
@arielj All issues should be addressed now. I see that one check has failed, but it was during installation of `node-gyp` so it's not related to this PR

imLinguin commented on src/frontend/screens/Game/GamePage/index.scss:
```diff
@@ -428,12 +428,43 @@
-  height: 100%;
-  gap: min(3vw, 50px);
+  display: grid;
+  position: relative;
```
Now that we have a `position: relative;` here, some images are slightly offset to the right, because background image is `width: 100vw;`. I'd suggest changing this to 100% in order to keep it centered

(disabled filters for better visibility in the example)

100vw
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/62100117/bc1bdf86-c433-4203-84eb-d5fb07df6cc4)

100%
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/62100117/f19b2585-1e19-4d3c-a617-2bf410208b8a)



Whole image:
![image](https://images.gog.com/62d0f061627c05c819adf950ac391c4aae76bacf121ee2dd1149d1cd9fe237a0.webp?namespace=gamesdb)

Kajot-dev replied:
I will change it to `100%` for both width and height since when viewport is <= 600px, then 100vh <= 100% and the image does not cover the whole background

Kajot-dev replied:
Fixed!

arielj reviewed (changes requested):
I was testing this, I found only one issue, now that the 3 dots button is back on the side of the screen, the menu open in the wrong direction for RTL languages:

![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/188464/eebaed67-0ca7-450b-9542-a56dde4090ac)

The old design handles the direction this opens for reference (if I remember right it's some css based on the RTL class).

then there's one more issue related to RTL languages, the `tabs` MUI component seems to not support it properly and the arrow are not the right icons and one of the icons never shows up, but I don't know if this is fixable if it's an issue with MUI itself

other than those 2 issues it looks great!

Kajot-dev commented:
> I was testing this, I found these 2 issues:
> 
> now that the 3 dots button is back on the side of the screen, the menu open in the wrong direction for RTL languages:
> 
> ![image](https://private-user-images.githubusercontent.com/188464/316677552-eebaed67-0ca7-450b-9542-a56dde4090ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0MTAwNTEsIm5iZiI6MTcxMTQwOTc1MSwicGF0aCI6Ii8xODg0NjQvMzE2Njc3NTUyLWVlYmFlZDY3LTBjYTctNDUwYi05NTQyLWE1NmRkZTQwOTBhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNVQyMzM1NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Y2QwZWRjOTZhYTBhYzUxOTVmNzU1YTBhMmUwOWRkOWFjYTEwMzg2OWIzMDM0NWE1YjEzNjk5ZDg0ZmNhNTc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.oOcbQmEEaIp10Pk1kSI_zbkrOyzX430qCAKahDzudHM)
> 
> The old design handles the direction this opens for reference (if I remember right it's some css based on the RTL class).
> 
> then there's one more issue related to RTL languages, the `tabs` MUI component seems to not support it properly and the arrows are not the right icons and one of the icons never shows up, but I don't know if this is fixable if it's an issue with MUI itself
> 
> other than those 2 issues it looks great!

O I didn't even know, that Heroic supports RTL! As of MUI tabs this issue affects all MUI tabs in heroic so I would guess that it's a global MUI issue (I found this: https://github.com/mui/material-ui/issues/30207 ; it's closed but not resolved really :)). I will think of some wrapper for the `Tabs` component which would correct this

Kajot-dev commented:
> I was testing this, I found these 2 issues:
> 
> now that the 3 dots button is back on the side of the screen, the menu open in the wrong direction for RTL languages:
> 
> ![image](https://private-user-images.githubusercontent.com/188464/316677552-eebaed67-0ca7-450b-9542-a56dde4090ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0MTAwNTEsIm5iZiI6MTcxMTQwOTc1MSwicGF0aCI6Ii8xODg0NjQvMzE2Njc3NTUyLWVlYmFlZDY3LTBjYTctNDUwYi05NTQyLWE1NmRkZTQwOTBhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNVQyMzM1NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Y2QwZWRjOTZhYTBhYzUxOTVmNzU1YTBhMmUwOWRkOWFjYTEwMzg2OWIzMDM0NWE1YjEzNjk5ZDg0ZmNhNTc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.oOcbQmEEaIp10Pk1kSI_zbkrOyzX430qCAKahDzudHM)
> 
> The old design handles the direction this opens for reference (if I remember right it's some css based on the RTL class).
> 
> then there's one more issue related to RTL languages, the `tabs` MUI component seems to not support it properly and the arrows are not the right icons and one of the icons never shows up, but I don't know if this is fixable if it's an issue with MUI itself
> 
> other than those 2 issues it looks great!

Fixed dots menu with a simple CSS tweak.
Fixed all MUI tabs across Heroic (We need to use mui Theme with specified direction as mui has internal logic which depends on that). So what happened was that CSS was RTL but the MUI logic was expecting LTR.

Kajot-dev commented:
Build on macos-12 failed while detaching a drive, not related to the PR

Kajot-dev commented:
Fixed some more issues (and ofc all of these that you mentioned):
- Nav buttons now respect color scheme
- Background is scaled up by 10% to mitigate ugly edges on blurred image (this was not visible at first, since image was scaled to 100vw and now is scaled to 100%)
- Instead of image filter `brightness(30%)` it now uses `opacity` to blend with the background. This makes background darker on dark themes (text is light) and brighter on light themes (text is dark usually)
- Also styled progressbar using `LinearProgress` (since both sidebar and downloads tab are using it) - this also affects old design slightly
- Icons on the main button are now larger and icon is always on the left side (this affects old design slightly)

Now I think everything is working as it should

arielj reviewed (approved):
hey @Kajot-dev can you update the PR to have the latest code from main? just to be sure it's all good (there's no conflicts but shouldn't hurt to do that and I can merge this)

Kajot-dev commented:
@arielj Pulled changes from main. Failed e2e, I think its unrelated (fails on testing a webview and I cannot reproduce the error), but you can rerun a Github job just in case

arielj commented:
thanks!
